crate_universe/bootstrap.bzl

```diff
@@ -0,0 +1,96 @@
+"""A module for declaraing a repository for bootstrapping crate_universe"""
+
+load("//crate_universe/private:util.bzl", "get_host_info")
+
+_INSTALL_SCRIPT_CONTENT = """\
+#!/bin/bash
+
+set -euo pipefail
+
+cp "${CRATE_RESOLVER_BIN}" "$@"
+"""
+
+_BUILD_FILE_CONTENT = """\
+package(default_visibility = ["//visibility:public"])
+
+exports_files(["release/crate_universe_resolver{ext}"])
+
+sh_binary(
+    name = "install",
+    data = [
+        ":release/crate_universe_resolver{ext}",
+    ],
+    env = {{
+        "CRATE_RESOLVER_BIN": "$(execpath :release/crate_universe_resolver{ext})",
+    }},
+    srcs = ["install.sh"],
+)
+"""
+
+def _crate_universe_resolver_bootstrapping_impl(repository_ctx):
+    resolver_triple, toolchain_repo, extension = get_host_info(repository_ctx)
+
+    cargo_path = repository_ctx.path(Label(toolchain_repo + "//:bin/cargo" + extension))
+    rustc_path = repository_ctx.path(Label(toolchain_repo + "//:bin/rustc" + extension))
+
+    repository_dir = repository_ctx.path(".")
+    resolver_path = repository_ctx.path("release/crate_universe_resolver" + extension)
+
+    args = [
+        cargo_path,
+        "build",
+        "--release",
+        "--locked",
+        "--target-dir",
+        repository_dir,
+        "--manifest-path",
+        repository_ctx.path(repository_ctx.attr.manifest),
+    ]
+
+    repository_ctx.report_progress("bootstrapping crate_universe_resolver")
+    result = repository_ctx.execute(
+        args,
+        environment = {
+            "RUSTC": str(rustc_path),
+            "RUST_LOG": "info",
+        },
+        quiet = False,
+    )
+
+    if result.return_code != 0:
+        fail("exit_code: {}".format(
+            result.return_code,
+        ))
+
+    repository_ctx.file("install.sh", _INSTALL_SCRIPT_CONTENT)
+
+    repository_ctx.file("BUILD.bazel", _BUILD_FILE_CONTENT.format(
+        ext = extension,
+    ))
+
+_crate_universe_resolver_bootstrapping = repository_rule(
+    doc = "A rule for bootstrapping a crate_universe_resolver binary",
+    implementation = _crate_universe_resolver_bootstrapping_impl,
+    attrs = {
+        "lockfile": attr.label(
```

Comment:
I'd be tempted to call this `cargo_lockfile` as it conflicts with the current crate_universe concept of a `lockfile`.

Reply:
I feel given the context (doc and attribute requirements) that it's clear what should go here. I can update it if you feel strongly but I'd like to avoid redundancy if possible.

Reply:
I feel pretty strongly about it - in particular, I can imagine a future bootstrapping world where we used `rules_rust` to build the resolver, and where we actually did use a `crate_universe` lockfile to do so. I don't think the naming/context is sufficiently uniquely disambiguating :)

Reply:
Updated